system_tests/recreatestate_rpc_test.go

```diff
@@ -459,3 +486,206 @@ func TestSkippingSavingStateAndRecreatingAfterRestart(t *testing.T) {
+	execConfig.Caching.BlockAge = 0      // use only Caching.BlockCount to keep only last N blocks in dirties cache, no matter how new they are
+	execConfig.Sequencer.MaxBlockSpeed = 0
+	execConfig.Sequencer.MaxTxDataSize = 150 // 1 test tx ~= 110
+	builder, cancelNode := prepareNodeWithHistory(t, ctx, execConfig, 16)
+	execNode, _ := builder.L2.ExecNode, builder.L2.Client
+	defer cancelNode()
+	bc := execNode.Backend.ArbInterface().BlockChain()
+	api := execNode.Backend.APIBackend()
+
+	header := bc.CurrentBlock()
+	if header == nil {
+		Fatal(t, "failed to get current block header")
+	}
+	state, _, err := api.StateAndHeaderByNumber(ctx, rpc.BlockNumber(header.Number.Uint64()))
+	Require(t, err)
+	addr := builder.L2Info.GetAddress("User2")
+	exists := state.Exist(addr)
+	err = state.Error()
+	Require(t, err)
+	if !exists {
+		Fatal(t, "User2 address does not exist in the state")
+	}
+	// Get the state again to avoid caching
+	state, _, err = api.StateAndHeaderByNumber(ctx, rpc.BlockNumber(header.Number.Uint64()))
+	Require(t, err)
+
+	blockCountRequiredToFlushDirties := builder.execConfig.Caching.BlockCount
+	makeSomeTransfers(t, ctx, builder, blockCountRequiredToFlushDirties)
+
+	exists = state.Exist(addr)
+	err = state.Error()
+	Require(t, err)
+	if !exists {
+		Fatal(t, "User2 address does not exist in the state")
+	}
+}
+
+func TestStateAndHeaderForRecentBlock(t *testing.T) {
+	threads := 32
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	builder := NewNodeBuilder(ctx).DefaultConfig(t, true)
+	builder.execConfig.Caching.Archive = true
+	builder.execConfig.RPC.MaxRecreateStateDepth = 0
+	cleanup := builder.Build(t)
+	defer cleanup()
+	builder.L2Info.GenerateAccount("User2")
+
+	errors := make(chan error, threads+1)
+	senderDone := make(chan struct{})
+	go func() {
+		defer close(senderDone)
+		for ctx.Err() == nil {
+			tx := builder.L2Info.PrepareTx("Owner", "User2", builder.L2Info.TransferGas, new(big.Int).Lsh(big.NewInt(1), 128), nil)
+			err := builder.L2.Client.SendTransaction(ctx, tx)
+			if ctx.Err() != nil {
+				return
+			}
+			if err != nil {
+				errors <- err
+				return
+			}
+			_, err = builder.L2.EnsureTxSucceeded(tx)
+			if ctx.Err() != nil {
+				return
+			}
+			if err != nil {
+				errors <- err
+				return
+			}
+			time.Sleep(10 * time.Millisecond)
+		}
+	}()
+	api := builder.L2.ExecNode.Backend.APIBackend()
+	db := builder.L2.ExecNode.Backend.ChainDb()
+	i := 1
+	var mtx sync.RWMutex
+	var wgCallers sync.WaitGroup
+	for j := 0; j < threads && ctx.Err() == nil; j++ {
+		wgCallers.Add(1)
+		go func() {
+			defer wgCallers.Done()
+			mtx.RLock()
+			blockNumber := i
+			mtx.RUnlock()
+			for blockNumber < 300 && ctx.Err() == nil {
+				prefix := make([]byte, 8)
+				binary.BigEndian.PutUint64(prefix, uint64(blockNumber))
+				prefix = append([]byte("b"), prefix...)
+				it := db.NewIterator(prefix, nil)
+				defer it.Release()
+				if it.Next() {
```

Comment:
suggestion: to speed up polling and hit more edge-cases, each thread could cache blockhashes that it already got a non-error response for in previous loops and skip them the next time iterator gets them

Reply:
this test is mostly intended to be a regression test for a specific issue, I added a comment describing the details in the follow up PR: https://github.com/OffchainLabs/nitro/pull/2442 

there should be only one key with the prefix (unless there was a reorg) so there is no need of caching. If the entry is found, the `recentBlock` (previously `i`) is advanced to the next block, and the iterator is created for the new prefix.

Let me know if you think that extending this test makes sense to cover more cases then the targeted issue